### ISSUE TITLE
API-methods to add/remove visual markers

### ISSUE THREAD

**Issue by hugbug** [lenmus/lomse#210]
This is a follow-up to #206.

It would be great to have easy to use methods to add/remove visual markings. In particular I need vertical lines to visually present selected portion of the score.

Possible definition:
 
```C++
enum MarkerType { VerticalLine, etc. };

virtual Object* add_visual_marker(ImoId scoreId, MarkerType markertype, TimeUnits timepos);

// maybe also alternatives for positioning, similar to "move_tempo_line":
virtual Object* add_visual_marker(ImoId scoreId, MarkerType markertype, int iMeasure, int iBeat, int iInstr=0);
virtual Object* add_visual_marker(ImoId scoreId, MarkerType markertype, int iMeasure, TimeUnits location, int iInstr=0);

virtual void remove_visual_marker(Object*);
```

The method should return an object which we can (after typecasting) adjust to our needs (change color, etc.). I'm not sure what type the object should have, I've typed `Object*` as a placeholder for now.

> marker heights (one staff, all saves for an instrument, all staves from instrument i to instrument j, other)

I don't know what is better here: extra parameters of `add_visual_marker` or can that be adjusted later via returned object's methods?

> it is necessary to consider if these markers have to be just visual effects, overlayed on the score, without representation in the document model or, by the contrary, these markers are real objects in the document model, such as the lines drawn in the 'drawlines' sample.

At least in my case I feel the markers should not be added to document as they are merely visual effects.

**Comment by cecilios** [comment]
Having though a litle about all this I think that there is a need for both types of markers:
- visual markers: just temporary objects to provide visual feedback to the user, and
- user markers: hand made annotations, annotations made by drawing tools, and annotations coded into the score, such as the current line objects, that it is necessary to preserve as part of the document. I'm thinking on things such us:
![user-marks-3](https://user-images.githubusercontent.com/5238679/64289035-c0827d80-cf63-11e9-8100-e7fc4a15c080.png)
![user-marks-5](https://user-images.githubusercontent.com/5238679/64289043-c5dfc800-cf63-11e9-9654-14fe912f8bc5.png)
![user-marks-6](https://user-images.githubusercontent.com/5238679/64289049-c9734f00-cf63-11e9-8280-8d752e00e2e4.png)

For your app. I also feel that visual markers are more appropiate. So, for now, let's focus on the visual markers.

All visual effects are part of the View, not part of the Document. Therefore they are owned by the View and must be created by the Interactor. Thus my proposal would be to add to the Interactor the following methods:

```
    /**
        Methods to add a marker. There are alternatives to facilitate the positioning
        of the marker.

        These methods return an ScoreMarker object whose visual properties (marker type,
         color, position, length, etc.) can later be changed to suit the user needs.

        When no value is specified for parameter iInstr (iInstr == -1) the marker
        will cover all staves in the system. If a value for iInstr is specified, the
        marker will cover only the staves of the specified instrument.
        The marker default height is the necessary to cover all staves of the selected
        range, from top staff line of fisrt staff to first line of last staff. An
        an extra length of three staff lines is added on top and bottom.

        Default properties are as follows:
        - Marker type is <tt>k_marker_line</tt>, that is, a vertical line.
        - Line color is transparent red (#ff000080).
        - Line thickness is 8 tenths.
        - Line style solid.
    */
    ScoreMarker* add_visual_marker(ImoId scoreId, TimeUnits timepos, int iInstr=-1);

    /** For multi-metric scores iMeasure and iBeat will refer to the specified
        instrument.
    */
    ScoreMarker* add_visual_marker(ImoId scoreId, int iMeasure, int iBeat, int iInstr=0);
    ScoreMarker* add_visual_marker(ImoId scoreId, int iMeasure, TimeUnits location,
                                   int iInstr=0);

    void remove_visual_marker(ScoreMarker*);


```
For now, the following markers will be considered:
```
enum EMarkerType
{
    k_marker_line = 0,          ///< Just a vertical line
    k_marker_open_squared,      ///< Squared bracket with horizontal lines at right
    k_marker_close_squared,     ///< Squared bracket with horizontal lines at left
    k_marker_open_curly,        ///< Curly bracket with central angle pointing to left
    k_marker_close_curly,       ///< Curly bracket with central angle pointing to right
};
```
![curly-brackets](https://user-images.githubusercontent.com/5238679/64289225-27a03200-cf64-11e9-83c1-f7cc2edd8d02.png)
![squared-brackets](https://user-images.githubusercontent.com/5238679/64289229-2bcc4f80-cf64-11e9-93dc-a2e5aba0d8bc.jpeg)

This is a code example of use:
```
        ImoId scoreId = ...
        TimeUnits timepos = ...
        ScoreMarker* marker = pInteractor->add_visual_marker(scoreId, timepos);

        //customize the marker. Will cover second and third instruments
        marker->color("#00ff00")->thickness(5)->top(1)->bottom(2);
            ...

        //reposition the marker and change its appearance
        marker->move_to(measure, beat);
        marker->type(k_marker_open_bracket)->color("ffff00");
            ...

        //when no longer needed remove it
        pInteractor->remove_visual_marker(marker);
```

The ScoreMarker will be a class derived from VisualEffect with the following interface:

```
class ScoreMarker : public VisualEffect
{
public:
        //customizable properties

    // marker height and position
    //iStaff == -1 means first staff for top or last staff for bottom
    ScoreMarker* x_shift(LUnits dx);
    ScoreMarker* top(int instr, int staff=-1, LUnits dy=0);
    ScoreMarker* bottom(int instr, int staff=-1, LUnits dy=0);

    /** Set the color of the marker.  */
    inline ScoreMarker* color(Color value) {...}
    inline ScoreMarker* color(const string& value) {...}

    /** Set the type of the marker.  */
    inline ScoreMarker* type(EMarkerType value) {...}

    /** Set the line thickness, in tenths.  */
    inline ScoreMarker* thickness(Tenths value) {...}

    /** Set the line style.  */
    inline ScoreMarker* line_style(ELineStyle value) {...}

        //access to current values

    /** Return the current marker color. */
    inline Color get_color() const {...}

    /** Return the current marker type. */
    inline EMarkerType get_type() const {...}
    
    ... etc.

    /** For multi-metric scores iMeasure and iBeat will refer to the specified
        instrument.
    */
    ScoreMarker* move_to(TimeUnits timepos);
    ScoreMarker* move_to(int iMeasure, int iBeat, int iInstr=0);
    ScoreMarker* move_to(int iMeasure, TimeUnits location, int iInstr=0);
};

```

Tomorrow I will continue with a second thought on this before starting to code. Feedback is welcome!!

**Comment by hugbug** [comment]
That looks fantastic :star2:
I can't wish more :joy: (joy smiley)

---

BTW, you can activate syntax highlighting for code fragments by adding `C++` just after these three apostrophe-like characters, for example:
 ```C++
```C++
enum EMarkerType
{
    k_marker_line = 0,          ///< Just a vertical line
    k_marker_open_squared,      ///< Squared bracket with horizontal lines at right
    k_marker_close_squared,     ///< Squared bracket with horizontal lines at left
    k_marker_open_curly,        ///< Curly bracket with central angle pointing to left
    k_marker_close_curly,       ///< Curly bracket with central angle pointing to right
};
```

**Comment by cecilios** [comment]
> BTW, you can activate syntax highlighting for code fragments by adding C++ just after these three apostrophe-like characters

Didn't know. Thank you!

Now some feedback.

I have reviewed the specification, did some preliminary coding and started the tests. For now there are a few changes in the specs and one open question.

**1. Canges in the API**

_a) names_

The name ‘visual marker’ is perhaps too generic. In future other types of visual score markings could be needed. For example, a shaded rectangular area similar to a selection rectangle or an ellipse enclosing a group of objects. Therefore, I think it is better to rename the class and some of the methods.
 
 As the markings we have to implement now are brackets and vertical lines spanning one or several staves, to define start and end points on the score or on a score part, I've changed the class name from  `ScoreMarker` to `StartEndMark`.
 
 Also, the creation methods need to be renamed from `add_visual_marker()` to `add_start_end_mark()`. This gives room for other future methods to create new marks, such as `add_ellipse_mark()` or `add_box_mark()`.
 
 _b) parameters for positioning the marks_
 
 The initial spec. was based on providing a timepos, either directly or indirectly by measure, beat.
 
 But timepos is only meaningful for notes/rest. The other staff objects (barlines, clefs, etc.) don't have duration and thus its timepos is the same than that of the following note/rest. For instance, in following score:
 
![timepos-uncertainty](https://user-images.githubusercontent.com/5238679/64437346-6948ee00-d0c6-11e9-859b-98927d460ec4.png)

 Assuming first eight note (duration 32) is at time 1000, after it (at barline) timepos will be 1032. As barlines do not take duration, the next quarter note after the barline will also be located at timepos 1032. So two different physical locations (barline and quarter note) have the same timepos. The same occurs with the clef and the next eight note.
 
 There is no uncertainty in specifying a timepos for moving the tempo line because it always moves to note/rests positions. But for positioning the visual marks, timepos is not enough as they should be placed with reference to any staff object. Therefore, method `add_visual_marker(timepos)` will be splitted in three:
 
 ```C++
 add_start_end_mark_at_event(TimeUnits timepos)
 add_start_end_mark_at_first_obj(TimeUnits timepos)
 add_start_end_mark_at_obj(ImoStaffObj* pSO)
 ```

**2. Open question**

It was decided to consider two types of editorial markings:
- User markings, that is markings on the document that must be preserved, and
- App markings, temporary visual marks created by an application to give feedback about relevant issues, such as start-end points for playback, set of ojects selected, current notes being played, etc.

We agreed that user markings must be part of the document and that app markings are just visual effects not stored in the document.

But, should app markings survive a document modification? For instance, if the app. modifies the document (e.g. by inserting/deleting notes) what should Lomse do with the marks? My feeling is that they must be deleted when the document is updated.

Feedback and oppinions welcome!

**Comment by hugbug** [comment]
>a) names

>`add_start_end_mark(), add_ellipse_mark(), add_box_mark()`

`start_end` sounds confusing a little bit. What about such words as `region`, `section`, `fragment`, `clip`? I'll be happy with any name, of course.

> b) parameters for positioning the marks

I hope one of these methods will allow to put the line at position without any (visible) note like the tempo line can:
![64280103-b440f500-cf50-11e9-8129-caf283adcdb3](https://user-images.githubusercontent.com/3368402/64461346-3a9a3a00-d0fd-11e9-8b4e-b321b074c21b.png)

> But, should app markings survive a document modification?

In my particular case the document is never edited and my answer to the question: whatever is easier to implement ;)

**Comment by cecilios** [comment]
PR #211 is a firts implementation.
Example of use:
```C++
ImoId scoreId = pScore->get_id();

//red mark: first note after barline (measure 1, beat 0)
TimeUnits timepos = ScoreAlgorithms::get_timepos_for(pScore, 1, 0); //measure, beat
spInteractor->add_fragment_mark_at_note_rest(scoreId, timepos);

//blue mark: at barline
FragmentMark* mark = spInteractor->add_fragment_mark_at_barline(scoreId, timepos);
mark->color(Color(0,0,255,128));

//green mark: at staffobj (clef). From second instrument to last one
ScoreCursor cursor(pDoc, pScore);     //cursor points to clef
cursor.move_next();         //now points to key signature
cursor.move_next();     //time signatute
ImoStaffObj* pSO = dynamic_cast<ImoStaffObj*>(*(cursor));
mark = spInteractor->add_fragment_mark_at_staffobj(scoreId, pSO);
mark->color(Color(0,255,0,128));
mark->top(1);

//magenta mark: at interpolated timepos. Only first instrument
mark = spInteractor->add_fragment_mark_at_note_rest(scoreId, 32);
mark->color(Color(255,0,255,128));
mark->bottom(0);

//cyan mark: instruments 2,3 & 4
timepos = ScoreAlgorithms::get_timepos_for(pScore, 1, 1);
mark = spInteractor->add_fragment_mark_at_note_rest(scoreId, timepos);
mark->color(Color(0,255,255,128));    //
mark->top(1,0);
mark->bottom(3);

//spInteractor->remove_mark(mark);

```

Result:
![fragment-marks-1](https://user-images.githubusercontent.com/5238679/64636264-a3d1c400-d401-11e9-9ade-cc91c11a35ec.png)

**Comment by hugbug** [comment]
Thank you very much. The API is beautifully designed and easy to use.
I’ll integrate this into my app ASAP (most likely tomorrow) and provide feedback.

**Comment by hugbug** [comment]
Works great for my needs:

![Screen Shot 2019-09-11 at 13 23 52](https://user-images.githubusercontent.com/3368402/64694678-81d85000-d49a-11e9-81de-5e0b7e36c828.png)

I'm also using `x_shift` to move the mark slightly to not overlap with tempo line:

![Screen Shot 2019-09-11 at 13 23 32](https://user-images.githubusercontent.com/3368402/64694725-a3d1d280-d49a-11e9-9e06-c41533417d42.png)

---

I'm observing a behaviour which I'm not sure is correct. In my app the score is filled into app window and when the window is resized I do:
```C++
pageInfo->set_page_width(ScaledUnits(m_image->getWidth()));
pageInfo->set_page_height(ScaledUnits(m_image->getHeight()));

interactor->on_document_updated();  //This rebuilds GraphicModel and will generate a Paint event
interactor->redraw_bitmap();
```

The score is reformatted correctly, the tempo line is displayed correctly too. The marks however remain on their old absolute screen positions. I've solved this by resetting marks after window resizing.

**Comment by cecilios** [comment]
> The score is reformatted correctly, the tempo line is displayed correctly too. The marks however remain on their old absolute screen positions. I've solved this by resetting marks after window resizing.

This is a side effect of the life-time span of user marks. At the end of this post  https://github.com/lenmus/lomse/issues/210#issuecomment-528887277 , I asked myself: _"should app markings survive a document modification?"_  and I decided that it is non-sense to keep visual effects alive. This is because when the document is updated and the graphic model is rebuild nothing ensures that existing marks are still valid, as pointed objects or pointed positions could no longer exist.

Therefore, all visual marks should be deleted when the document is modified, and user application should re-create the marks if still needed. But this creates new issues. There are two options (if you have other ideas please share):

a) Current situation: Lomse does not delete the marks and thus, the user application will the notice this _bad mark position behaviour_ when the graphic model is updated. But this will be noticed during the development application test and fixed by deletening and recreating the marks if still needed.

b) Delete all user marks automatically when then graphic model is rebuild. In this case, if the user application tries to use the pointers to the marks, this will cause crashes as the marks no longer exists. One solution is to return shared pointers to the marks instead of raw pointers, for preventing using the pointer when the object has been deleted.

For now, while waiting for your tests results and your feedback, I did not deleted the marks when the model is rebuild.

**Comment by cecilios** [comment]
After writing the previous post I think there is a third option:

c) To store the FragmentMark creation parameters adn then, after a document update, Lomse will be able to re-position the marks if the timepos/satffobj still exist. But if the time position or the reference staffobj no longer exist, what to do? Deleting the mark could be the best answer but this again, requires to use smart pointers just in case the user application tries to use the deleted marks.

Perhaps this is the most robust solution that simplifies the life of people using the Lomse library.

**Comment by cecilios** [comment]
By the way, your code
```C++
pageInfo->set_page_width(ScaledUnits(m_image->getWidth()));
pageInfo->set_page_height(ScaledUnits(m_image->getHeight()));
```
althougth correct, suggest me the possible need for some mising API function. What are you trying to achieve by adjusting score page to window size? What type of view are you using?

**Comment by hugbug** [comment]
I haven’t realized that changing page size is a document modification but now it makes sense to me. Recreating marks is an easy task. I suggest to keep the things as they are now.

- - -
> What are you trying to achieve by adjusting score page to window size?

This is to ensure the user never needs to scroll horizontally. When the app window is resized the score view takes all available width. I’m using `k_view_vertical_book`.
![Playback with scores](https://user-images.githubusercontent.com/3368402/64079783-1b11a480-ccec-11e9-9511-b94286b3855a.png)

```C++
LUnits LomseScoreComponent::ScaledUnits(int pixels)
{
	return LUnits(pixels) * 26.5f / m_scale;
}
```
`m_scale` is current scale factor of program UI (1 by default). The scaling is provided by UI library.
 I’m not sure where the constant `26.5` comes from.

**Comment by cecilios** [comment]
A really nice UI!!

> This is to ensure the user never needs to scroll horizontally. When the app window is resized the score view takes all available width.

Ok, thank you. Your approach is right and for now I cannot think about alterantive API methods.

As to the 26.5 factor in your `ScaledUnits()` formula I deduce that it comes from considering a fixed screen resolution of 96 dpi.

If P is the view width (pixels), and R is the screen resolution (dpi) then:

	L (view width in LUnits --cents of mm--) = (P/R) * 25.4 * 100

Considering a resolution of 96 dpi we get:

	L = (P * 25400)/96 = P * 26.4583333333333  that is, approx. L = P * 26.5

You should investigate if the value m_scale provided by the UI library takes screen resolution into account for consistency between different platforms. Otherwise, perhaps you should consider the screen resolution in your `ScaledUnits()` formula.

**Comment by cecilios** [comment]
PR #212 closes this issue. 

I have updated the API documentation. Please see:
- http://lenmus.github.io/lomse/classFragmentMark.html
- http://lenmus.github.io/lomse/classInteractor.html#aeafc7bcddce787cdfb9769feed5b8b63

I'm going to close this issue. Feel free to continue using this issue to If you have questions or to document ideas and findings in relation to fragment markers. I can re-open it later if changes are necessary.

**Comment by hugbug** [comment]
It's very nice to have different types of marks. I'm using this combination for now:

![Screen Shot 2019-09-15 at 22 30 14](https://user-images.githubusercontent.com/3368402/64927258-a1f47000-d808-11e9-9633-de81a66143d0.png)

---

Discovered a minor issue with the current version. After resizing (which causes `interactor->on_document_updated(); interactor->redraw_bitmap()`) the tempo line disappears and needs to be redrawn with `move_tempo_line`. This didn't happen with previous version.

**Comment by cecilios** [comment]
> After resizing (which causes interactor->on_document_updated(); interactor->redraw_bitmap()) ...

It is not necessary to invoke `interactor->on_document_updated()`. Invoking `interactor->redraw_bitmap()` should be enough because it checks if the document has been updated and, if true, rebuilds the graphic model. But invoking `interactor->on_document_updated()` does neither causes harm nor more workload.

> Discovered a minor issue with the current version. [...] This didn't happen with previous version.

Please re-check this! Whenever the graphic model is deleted and rebuild, all visual effects owned by the GraphicView (the tempo line among others), are removed. Therefore, when you invoke `interactor->on_document_updated()` the tempo line is removed. This has been always the behaviour and nothing has been changed in relation to this. In fact, the decision about deleting the fragment marks is still pending and, so, the code for removing visual marks when the graphic model is rebuild has not been touched. I've checked the sources and the changes, and previous version also deleted the tempo line!

Another issue: After updating the API documentation I noticed the need for a couple of changes. With reference to the picture in the API docs (reproduced here):
![FragmentMark-types](https://user-images.githubusercontent.com/5238679/64945729-84a7bc00-d871-11e9-8469-42780ba55e8a.png)

- When positioning the marks on barlines the marks are centered when using k_mark_line but not with the others. I'm going to change this so method `Interactor::add_fragment_mark_at_barline()` will force all mark types to be centered.

- Marks covering the same instruments have differente heights, because bracket hooks are added on top and bottom of the line height. I feel that all marks sould have the same height (equal to that of k_mark_line); otherwise the user will have less control about the exact heigh of the marks.

**Comment by hugbug** [comment]
>Please re-check this! Whenever the graphic model is deleted and rebuild, all visual effects owned by the GraphicView (the tempo line among others), are removed.

Tested with the previous version and the tempo line was indeed removed. Sorry for false alarm.

>It is not necessary to invoke interactor->on_document_updated(). Invoking interactor->redraw_bitmap() should be enough because it checks if the document has been updated and, if true, rebuilds the graphic model.

On window resizing I'm doing:
```C++
	pageInfo->set_page_width(ScaledUnits(m_image->getWidth()));
	pageInfo->set_page_height(ScaledUnits(m_image->getHeight()));

	interactor->on_document_updated();

	UpdateABMarks();
	UpdateTempoLine();

	interactor->redraw_bitmap();
```

If I remove call to `interactor->on_document_updated()` the changes to page size (`pageInfo->set_page_width`) are not updated and the document is rendered with old page width.

If you fix this and `redraw_bitmap` will know about necessity to update graphic model after changes of page properties - where would I put my `UpdateABMarks` and `UpdateTempoLine` calls? ;) Markers and tempo line need to be put after the graphic model is updated and before the bitmap is generated.

Being able to insert custom code between `on_document_updated` and `redraw_bitmap` seems to be the way to go.

**Comment by cecilios** [comment]
> Tested with the previous version and the tempo line was indeed removed. Sorry for false alarm.

Glad to know! It would have been a terrible issue otherwise!

> If you fix this and redraw_bitmap will know about necessity to update graphic model after changes of page properties - where would I put my UpdateABMarks and UpdateTempoLine calls? ;) Markers and tempo line need to be put after the graphic model is updated and before the bitmap is generated.

You are totally rigth and your code is correct. You need to re-create the graphic model before placing the markers and so your code is correct. I thought that you were invoking `interactor->on_document_updated()` inmediatelly followed by `interactor->redraw_bitmap()` and thus, I thought that that was not necessary althought no extra work is done by Lomse. But, I didn't realize that in your case `interactor->redraw_bitmap() will not detect that the document has been modified. This is because marking the document as *dirty* is only done in API oriented methods but not in most low level methods for modifying ImoObj attributes. 

In summary: After direct modification of ImoObj attributes the document is not market as dirty. Therefore it is necessary to delete the graphic model and recreate it:

```
interactor->on_document_updated();		//delete graphic model and re-builds it
//insert here code for updating visual effects, if needed
interactor->redraw_bitmap();			//triggers the repaint
```
